### DISCUSSION TITLE
How to best use DataStores for writing out sequence data?

### DISCUSSION THREAD

**Discussion by KatherineCaley** [cogent3/cogent3#1484] [General]
I'm using composable cogent3 "apps" to align a collection of bacterial sequences. I am interested in the best way to write them out so that I can re-load the aligned sequences for further analysis. 

My alignment routine is straightforward:

``` python
from cogent3 import get_app
from cogent3.app.io import open_data_store

# I open up the DataStores for my input and output
# `gene_collections` is a directory of unaligned fasta files
in_dstore = open_data_store("gene_collections", suffix=".fa", mode="r")
out_dstore = open_data_store("aligned.sqlitedb", mode="w")

# I grab the apps that I need 
reader = get_app("load_unaligned", moltype="dna")
trim_stop = get_app("trim_stop_codons")
codon_aligner = get_app("progressive_align", "codon")
writer = get_app("write_seqs", out_dstore)

# create a process composed of the apps 
process = reader + trim_stop + codon_aligner + writer

# I apply the process to the data
r = process.apply_to(
    dstore,
    show_progress=True,
    parallel=True,
)
```
It's not clear how to load the alignment stored in the SQLite DB. 

My approach is as follows: 

```python
reader = get_app("load_aligned")
dstore = open_data_store("aligned.sqlitedb", mode="r") 
alns = reader(store)
alns
# TypeError: DataStoreSqlite.read() missing 1 required positional argument: 'identifier'
```

This also happens when using `load_db` to load or `write_db` to write the alignments out. 

Have I written out the alignment in the incorrect way, in which case, this could use some extra documentation, or is this a bug, in which case I can migrate this post to an issue?

**Comment by KatherineCaley** [comment]
Note that my current workaround is to define a custom writer function using `define_app`. I'm just curious as to how this can be done using the pre-defined apps.

**Reply by GavinHuttley** [reply]
```python
reader = get_app("load_aligned")
dstore = open_data_store("aligned.sqlitedb", mode="r") 
alns = reader(dstore[0]) # note that a DataStore is a collection of records, so [0] is the first one
alns
```
To load them all, I *think* the following will work
```python
alns = list(reader.as_completed(dstore))
```

**Reply by KatherineCaley** [reply]
Brilliant. Both work! Thank you!
